lib/Frontend/ArgsToFrontendOutputsConverter.cpp

```diff
@@ -0,0 +1,205 @@
+  llvm::ErrorOr<std::unique_ptr<llvm::MemoryBuffer>> buffer =
+      llvm::MemoryBuffer::getFile(filelistPath);
+  if (!buffer) {
+    diags.diagnose(SourceLoc(), diag::cannot_open_file, filelistPath,
+                   buffer.getError().message());
+    return None;
+  }
+  std::vector<std::string> outputFiles;
+  for (StringRef line : make_range(llvm::line_iterator(*buffer.get()), {})) {
+    outputFiles.push_back(line.str());
+  }
+  return outputFiles;
+}
+
+Optional<std::vector<std::string>>
+OutputFilesComputer::getOutputFilenamesFromCommandLineOrFilelist(
+    const ArgList &args, DiagnosticEngine &diags) {
+  if (const Arg *A = args.getLastArg(options::OPT_output_filelist)) {
+    assert(!args.hasArg(options::OPT_o) &&
+           "don't use -o with -output-filelist");
+    return ArgsToFrontendOutputsConverter::readOutputFileList(A->getValue(),
+                                                              diags);
+  }
+  return args.getAllArgValues(options::OPT_o);
+}
+
+Optional<OutputFilesComputer>
+OutputFilesComputer::create(const llvm::opt::ArgList &args,
+                            DiagnosticEngine &diags,
+                            const FrontendInputsAndOutputs &inputsAndOutputs) {
+  Optional<std::vector<std::string>> outputArguments =
+      getOutputFilenamesFromCommandLineOrFilelist(args, diags);
+  if (!outputArguments)
+    return None;
+  const StringRef outputDirectoryArgument =
+      outputArguments->size() == 1 &&
+              llvm::sys::fs::is_directory(outputArguments->front())
+          ? StringRef(outputArguments->front())
+          : StringRef();
+  ArrayRef<std::string> outputFileArguments =
+      outputDirectoryArgument.empty() ? ArrayRef<std::string>(*outputArguments)
+                                      : ArrayRef<std::string>();
+  const StringRef firstInput = inputsAndOutputs.hasSingleInput()
+                                   ? inputsAndOutputs.getFilenameOfFirstInput()
+                                   : StringRef();
+  const FrontendOptions::ActionType requestedAction =
+      ArgsToFrontendOptionsConverter::determineRequestedAction(args);
+
+  if (!outputFileArguments.empty() &&
+      outputFileArguments.size() !=
+          inputsAndOutputs.countOfInputsProducingMainOutputs()) {
+    diags.diagnose(
+        SourceLoc(),
+        diag::error_if_any_output_files_are_specified_they_all_must_be);
+    return None;
+  }
+
+  return OutputFilesComputer(
+      args, diags, inputsAndOutputs, outputFileArguments,
+      outputDirectoryArgument, firstInput, requestedAction,
+      args.getLastArg(options::OPT_module_name),
+      FrontendOptions::suffixForPrincipalOutputFileForAction(requestedAction),
+      FrontendOptions::doesActionProduceTextualOutput(requestedAction));
+}
+
+OutputFilesComputer::OutputFilesComputer(
+    const llvm::opt::ArgList &args, DiagnosticEngine &diags,
+    const FrontendInputsAndOutputs &inputsAndOutputs,
+    ArrayRef<std::string> outputFileArguments,
```

Comment:
Nitpick: This may as well take a std::vector. That would avoid a copy if you `std::move` at the call site.

Reply:
Fixed.